### PR TITLE
Type Inferencer for S3

### PR DESCRIPTION
**Pull request by raynoldng**
Add typing support for Source 3, which introduces `for`/`while` loops, reassignment (`let` statements) and arrays. 

Added `set_head` and `set_tail` to the list of predefined functions. 

Updated specs, pdf available here: 
[source_3_type_inference.pdf](https://github.com/source-academy/js-slang/files/4558166/source_3_type_inference.pdf)

# Limitations

As we require lists to be monomorphic, the type inferencer breaks on trees and streams programs. See test cases: 2.2.2 (trees) and 3.5.1 (streams)

### CODE REVIEW

**Comment by martin-henz** [comment]
Great. FYI: Daryl is working on the specs for S2, so there would be a hand-over.

**Comment by martin-henz** [comment]
Also note the merge conflicts. Please resolve.

File: src/errors/typeErrors.ts [hunk 1]
```diff
@@ -1,11 +1,111 @@
...
+  ) {}
+
+  get location() {
+    return this.node.loc!
+  }
+
+  public explain() {
+    const [varName, assignmentStr] = formatAssignment(this.node)
+    return stripIndent`
+    Expected reassignment of ${varName}:
```
Comment (martin-henz) [inline src/errors/typeErrors.ts:89]: How about:
Expected assignment of ${varName}:
...
to get a value of type:
...
but got a value of type:
...

Comment (raynoldng) [reply src/errors/typeErrors.ts:89]: Yes, that is easier to read. Thanks for the suggestion!

Comment (martin-henz) [reply src/errors/typeErrors.ts:89]: I suggested the use of the term "assignment" and not "reassignment". Any assignment is a "reassignment", so the "re" doesn't add anything.

File: src/typeChecker/__tests__/typeChecker.test.ts [hunk 2]
```diff
@@ -579,10 +600,485 @@ describe('type checking functions used in polymorphic fashion', () => {
...
+    `
+    const [program, errors] = parseAndTypeCheck(code, 3)
+    expect(topLevelTypesToString(program)).toMatchInlineSnapshot(`
+      "arr: Array<number>
+      y: number
+      x: number"
+    `)
+    expect(parseError(errors)).toMatchInlineSnapshot(`
+      "Line 4: Array expected type: Array<number>
+          but got: boolean"
```
Comment (openorclose) [inline src/typeChecker/__tests__/typeChecker.test.ts:1001]: This error seems rather confusing.  Array<number> is the type of the expected array but boolean is the type of the expected array _element_ 

Ideally we should be consistent and display either the array type of both or the element type of both

**Pull request merged**